fili-core/src/main/java/com/yahoo/bard/webservice/sql/DruidQueryToSqlConverter.java

```diff
@@ -0,0 +1,444 @@
+     * @param druidQuery  The query to convert to sql.
+     * @param apiToFieldMapper  The mapping between api and physical names for the query.
+     *
+     * @return the sql equivalent of the query.
+     *
+     * @throws SQLException if can't connect to database.
+     */
+    public String buildSqlQuery(
+            Connection connection,
+            DruidAggregationQuery<?> druidQuery,
+            ApiToFieldMapper apiToFieldMapper
+    )
+            throws SQLException {
+        String sqlTableName = druidQuery.getDataSource().getPhysicalTable().getName();
+        String timestampColumn = DatabaseHelper.getTimestampColumn(
+                connection,
+                calciteHelper.getSchemaName(),
+                sqlTableName
+        );
+
+        LOG.info("Using timestamp column of '{}' for table {}", timestampColumn, sqlTableName);
+
+        RelBuilder builder = calciteHelper.getNewRelBuilder();
+        builder.scan(sqlTableName)
+                .project(
+                        getColumnsToSelect(builder, druidQuery, timestampColumn, apiToFieldMapper)
+                );
+        RelNode rootRelNode = builder.build();
+        RelToSqlConverter relToSql = calciteHelper.getNewRelToSqlConverter();
+        SqlPrettyWriter sqlWriter = calciteHelper.getNewSqlWriter();
+        return getIntervals(druidQuery)
+                .map(interval -> {
+                    RelBuilder localBuilder = calciteHelper.getNewRelBuilder();
+
+                    localBuilder.push(rootRelNode)
+                            .filter(
+                                    getAllWhereFilters(
+                                            localBuilder,
+                                            druidQuery,
+                                            timestampColumn,
+                                            interval
+                                    )
+                            )
+                            .aggregate(
+                                    localBuilder.groupKey(
+                                            getAllGroupByColumns(localBuilder, druidQuery, timestampColumn)
+                                    ),
+                                    getAllQueryAggregations(localBuilder, druidQuery, apiToFieldMapper)
+                            )
+                            .filter(
+                                    getHavingFilter(localBuilder, druidQuery, apiToFieldMapper)
+                            )
+                            .sortLimit(
+                                    NONE,
+                                    getThreshold(druidQuery),
+                                    getSort(localBuilder, druidQuery, apiToFieldMapper)
+                            );
+                    String sql = writeSql(sqlWriter, relToSql, localBuilder);
+                    return "(" + sql + ")";
+                })
+                .collect(Collectors.joining("\nUNION ALL\n"));
```

Comment:
Ok so after looking into it further, it appears this is supported by MySQL/PostgreSQL (not sure what else), but is not supported by Standard SQL. So for example SQLite will fail with this method.

[Relevant Stackoverflow](https://stackoverflow.com/questions/4653124/what-does-the-sql-standard-say-about-parentheses-in-sql-union-except-intersect-s)

So it looks like the best way to accomplish the `ORDER BY` and `LIMIT` functionality as subqueries (these are used by TopN over each bucket and the results are unioned together) would be to make multiple queries and aggregate the results manually like @michael-mclawhorn suggested.

I've filed a [bug report with Calcite](https://issues.apache.org/jira/browse/CALCITE-1892), but since I'm not sure how much can be done for this issue.